lib/core/include/experimental_plugin_framework.hpp

```diff
@@ -0,0 +1,461 @@
+#ifndef EXPERIMENTAL_PLUGIN_FRAMEWORK
+#define EXPERIMENTAL_PLUGIN_FRAMEWORK
+
+// =-=-=-=-=-=-=-
+// irods includes
+#include "rodsDef.h"
+#include "msParam.h"
+#include "rcConnect.h"
+#include "irods_plugin_base.hpp"
+
+#include "thread_pool.hpp"
+#include "connection_pool.hpp"
+#include "dispatch_processor.hpp"
+
+#ifdef RODS_SERVER
+#undef RODS_SERVER
+#define REDEF_RODS_SERVER
+#endif
+#include "query_builder.hpp"
+
+#ifdef REDEF_RODS_SERVER
+#define RODS_SERVER
+#endif
+
+// =-=-=-=-=-=-=-
+// stl includes
+#include <sstream>
+#include <string>
+#include <iostream>
+#include <termios.h>
+#include <unistd.h>
+
+#include "json.hpp"
+using json = nlohmann::json;
+
+namespace irods::experimental::api {
+    using flag_type = std::atomic_bool;
+
+    namespace states {
+        const std::string paused{"paused"};
+        const std::string failed{"failed"};
+        const std::string running{"running"};
+        const std::string unknown{"unknown"};
+        const std::string complete{"complete"};
+    }; // states
+
+    namespace commands {
+        const std::string pause{"pause"};
+        const std::string resume{"resume"};
+        const std::string cancel{"cancel"};
+        const std::string request{"request"};
+        const std::string progress{"progress"};
+    }; // commands
+
+    namespace endpoints {
+        const std::string command{"command"};
+        const std::string operation{"operation"};
+    }; // endpoints
+
+    namespace constants {
+        const std::string code{"code"};
+        const std::string status{"status"};
+        const std::string plugin{"plugin"};
+        const std::string errors{"errors"};
+        const std::string command{"command"};
+        const std::string message{"message"};
+        const std::string progress{"progress"};
+    }; // constants
+
+    class base;
+
+    namespace {
+        void sleep()
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(500));
+        }
+
+
+
+        auto invoke(rcComm_t& comm, const json& msg)
+        {
+            const uint32_t ADAPTOR_APN{120000};
```

Comment:
does this `120000` need to be documented/declared/coordinated anywhere else?

Reply:
and adapter with an e, rather than adaptor?

Reply:
the API number is hidden behind the client side interface everyone is expected to call to invoke a new API plugin